docs/bots-guide.md

```diff
@@ -443,6 +443,103 @@ bot_handler.update_message(dict(
+ (the actual test is written slightly more compact).
+
+     #!/usr/bin/env python
+
+    from __future__ import absolute_import
+    from __future__ import print_function
+
+    from six.moves import zip
+
+    from zulip_bots.test_lib import BotTestCase  # A base class for bot test suites that comes with various helpful testing methods.
+
+    class TestHelloWorldBot(BotTestCase):  # Our test suite for the bot.
+        bot_name = "helloworld"  # The bot's name (equivalent to the name of the bot module).
+
+        def test_bot(self): # A test case (must start with `test`)
+            # Messages we want to test and the expected bot responses.
+            message_response_pairs = {"" : "beep boop",
+                                      "foo" : "beep boop",
+                                      "Hi, my name is abc" : "beep boop"}
+            self.check_expected_responses(message_response_pairs)  # Test the bot with our message_response_pair dict.
+
+The `helloworld` bot replies with "beep boop" to every message @-mentioning it. Note that our helper method
+`check_expected_responses` adds the @-mention for us  - the only thing we need to do is to specify the
+rest of the message and the expected response. In this case, we want to assert that the bot always replies with "beep boop".
+To do so, we specify several test messages ("", "foo", "Hi, my name is abc") and assert that the response is always "beep boop".
+
+#### Test your test
+
+Once you have written a test suite, you want to verify that everything works as expected.
+
+* To test a bot in [Zulip's bot directory](https://github.com/zulip/python-zulip-api/tree/master/zulip_bots/zulip_bots/bots):
+  `tools/test-bots <botname>`
+
+* To run any test: `python -m unittest -v <package.bot_test>`
+
+* To run all bot tests: `tools/test-bots`
+
+This setup is sufficient for all bots without special requirements and features.
+
+#### Advanced testing
+
+This section shows advanced testing techniques for bots with special requirements or features.
+*The code for the bot testing library can be found [here](
+ https://github.com/zulip/python-zulip-api/blob/master/zulip_bots/zulip_bots/test_lib.py).*
+
+##### Asserting individual messages
+
+    self.assert_bot_response(
+        message = {'content': 'foo'},
+        response = {'content': 'bar'},
+        expected_method='send_reply'
+    )
+
+Use `assert_bot_response()` to test individual messages. Specify additional message settings, such as the stream or subject, in the
+`message` and `response` dicts.
+
+##### Testing bots with config files
+
+    with self.mock_config_info({'entry': 'value'}):
+        # self.assert_bot_response(...)
+
+`mock_config_info()` mocks a bot's config file. All config files are specified in the .ini format, with one
+default section. The dict passed to `mock_config_info()` specifies the keys and values of that section.
+
+##### Testing bots with internet access
+
```

Comment:
Again, explain what this is for.  "Some bots, such as Giphy, depend on a third-party we service, such as the Giphy webapp, in order to work.  Because we want our test suite to be reliable and not add load to these third-party APIs, tests for these services need to have "test fixtures": sample HTTP request/response pairs to be used by the tests.
You can specify which one to use in your test code using the following code:"